relay-server/src/actors/processor.rs

```diff
@@ -1449,21 +1450,29 @@ impl EnvelopeProcessorService {
         enable_cardinality_limiter: bool,
         _organization_id: u64,
         buckets: Vec<Bucket>,
-    ) -> Result<Box<dyn Iterator<Item = Bucket>>, relay_cardinality::Error> {
+    ) -> impl Iterator<Item = Bucket> {
         if !enable_cardinality_limiter {
-            return Ok(Box::new(buckets.into_iter()));
+            return Either::Left(buckets.into_iter());
         }
 
         #[cfg(feature = "processing")]
         if let Some(ref cardinality_limiter) = self.inner.cardinality_limiter {
-            return Ok(Box::new(
-                cardinality_limiter
-                    .check_cardinality_limits(_organization_id, buckets)?
-                    .into_accepted(),
-            ));
+            let limits = cardinality_limiter.check_cardinality_limits(_organization_id, buckets);
+
+            return match limits {
+                Ok(limits) => Either::Right(limits.into_accepted()),
+                Err((buckets, error)) => {
+                    relay_log::error!(
+                        error = &error as &dyn std::error::Error,
+                        "cardinality limiter failed"
+                    );
+
+                    Either::Left(buckets.into_iter())
+                }
+            };
         }
 
-        Ok(Box::new(buckets.into_iter()))
+        Either::<_, relay_cardinality::limiter::Accepted<_>>::Left(buckets.into_iter())
```

Comment:
Why does this need type annotations?

Reply:
For when we compile without `feature = processing`, then the `Right` side is never constructed -> error without annotations